src/Uno.UI.Runtime.Skia.Wpf/WPF/WpfDragDropExtension.cs

```diff
@@ -0,0 +1,279 @@
+
+		public WpfDragDropExtension(object owner)
+		{
+			_manager = (DragDropManager)owner;
+
+			var host = WpfHost.Current;
+
+			host.AllowDrop = true;
+
+			host.DragEnter += OnHostDragEnter;
+			host.DragOver += OnHostDragOver;
+			host.DragLeave += OnHostDragLeave;
+			host.Drop += OnHostDrop;
+		}
+
+		private void OnHostDragEnter(object sender, DragEventArgs e)
+		{
+			var src = new DragEventSource(_fakePointerId, e);
+			var data = ToDataPackage(e.Data);
+			var allowedOperations = ToDataPackageOperation(e.AllowedEffects);
+			var info = new CoreDragInfo(src, data.GetView(), allowedOperations);
+
+			CoreDragDropManager.GetForCurrentView()?.DragStarted(info);
+
+			// Note: No needs to _manager.ProcessMove, the DragStarted will actually have the same effect
+		}
+
+		private void OnHostDragOver(object sender, DragEventArgs e)
+			=> e.Effects = ToDropEffects(_manager.ProcessMoved(new DragEventSource(_fakePointerId, e)));
+
+		private void OnHostDragLeave(object sender, DragEventArgs e)
+			=> e.Effects = ToDropEffects(_manager.ProcessAborted(new DragEventSource(_fakePointerId, e)));
+
+		private void OnHostDrop(object sender, DragEventArgs e)
+			=> e.Effects = ToDropEffects(_manager.ProcessDropped(new DragEventSource(_fakePointerId, e)));
+
+		public void StartNativeDrag(CoreDragInfo info)
+			=> WpfHost.Current.Dispatcher.InvokeAsync(async () =>
+			{
+				try
+				{
+					var data = await ToDataObject(info.Data, CancellationToken.None);
+					var effects = ToDropEffects(info.AllowedOperations);
+
+					DragDrop.DoDragDrop(WpfHost.Current, data, effects);
+				}
+				catch (Exception e)
+				{
+					this.Log().Error("Failed to start native Drag and Drop.", e);
+				}
+			});
+
+		private static DataPackageOperation ToDataPackageOperation(DragDropEffects wpfOp)
+			=> (DataPackageOperation)((int)wpfOp) & (DataPackageOperation.Copy | DataPackageOperation.Move | DataPackageOperation.Link);
+
+		private static DragDropEffects ToDropEffects(DataPackageOperation uwpOp)
+			=> (DragDropEffects)((int)uwpOp) & (DragDropEffects.Copy | DragDropEffects.Move | DragDropEffects.Link);
+
+		private static DataPackage ToDataPackage(IDataObject src)
+		{
+			var dst = new DataPackage();
+			var text = src.GetData(DataFormats.Text) as string
+				?? src.GetData(DataFormats.UnicodeText) as string
+				?? src.GetData(DataFormats.OemText) as string
+				?? src.GetData(DataFormats.StringFormat) as string;
+
+			if (text.HasValueTrimmed())
+			{
+				if (Uri.IsWellFormedUriString(text, UriKind.Absolute))
+				{
+					dst.SetWebLink(new Uri(text!));
+				}
+				else
+				{
+					dst.SetText(text!);
+				}
+			}
+
+			if (src.GetData(DataFormats.Html) is string html)
+			{
+				dst.SetHtmlFormat(html);
+			}
+
+			if (src.GetData(DataFormats.Rtf) is string rtf)
+			{
+				dst.SetRtf(rtf);
+			}
+
+			if (src.GetData(DataFormats.Bitmap) is BitmapSource bitmap)
+			{
+				dst.SetBitmap(new RandomAccessStreamReference(async ct =>
```

Comment:
For large images this copy may cause an unwanted performance hit as the pointer first drags over an Uno app. Instead it might be best to use a DataProvider as commented below for StorageItems.

Reply:
It's very good point, however here I only create a `RandomAccesStream**Reference**`, which is designed to load its content only when [the reference is open for read](https://github.com/unoplatform/uno/blob/master/src/Uno.UWP/Storage/Streams/RandomAccessStreamReference.cs#L44). So it's also deferred to the effective usage.

I will however add some unit test to make sure that this behavior is not altered.

Reply:
Yes, I realized the RandomAccessStreamReference won't load the stream until it's needed. However, to create the stream reference you **are** making a copy which is what I was worried about:

```csharp
var copy = new MemoryStream();
var encoder = new BmpBitmapEncoder();
encoder.Frames.Add(BitmapFrame.Create(bitmap));
encoder.Save(copy);
copy.Position = 0;
```